### PR TITLE
Add modularized SSL training recipe

### PR DESCRIPTION
**Pull request by nateanl**
TorchAudio currently has one training recipe for HuBET + LibriSpeech pre-training. It may not suit well when users want to use customized dataset, or use a new training objective (such as contrastive loss in Wav2Vec2). The PR addresses the issue by providing a modularized training recipe for audio self-supervised learning. Users can inject customized model module, loss function, optimizer, lr scheduler, and datamodule for training a SSL model.

### CODE REVIEW

File: examples/ssl/train_hubert.py [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+        max_steps=args.max_updates,
+        num_nodes=args.num_nodes,
+        devices=args.gpus,
+        accelerator="gpu",
+        strategy="ddp",
+        replace_sampler_ddp=False,
+        callbacks=callbacks,
+        reload_dataloaders_every_n_epochs=1,
+    )
+    if args.model_name == "hubert_pretrain_base":
```
Comment (carolineechen) [inline examples/ssl/train_hubert.py:122]: can do assert args.model_name in [...] and then use getattr instead of multiple if statements

File: examples/ssl/losses/_hubert_loss.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import torch
+import torch.nn.functional as F
+from torch import Tensor
+
+
+def hubert_loss(
+    logit_m: Optional[Tensor],
+    logit_u: Optional[Tensor],
+    feature_penalty: Tensor,
+    label: Optional[Tensor] = None,
```
[the remark this answers is not in the document]
Comment (nateanl) [reply examples/ssl/losses/_hubert_loss.py:12]: yes, this is intended so that different tasks can use the same `loss = self.loss_fn(*out, *batch["label"])` in lightning.py

File: examples/ssl/losses/_hubert_loss.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    Args:
+        logit_m (Tensor or None): The masked logit Tensor of dimension `(masked_frames, final_dim)`.
+        logit_u (Tensor or None): The non-masked logit Tensor of dimension `(unmasked_frames, final_dim)`.
+        feature_penalty (Tensor): The feature mean value for additional penalty loss.
+        masked_weight (float, optional): The weight for masked cross-entropy loss (Default: ``1.0``).
+        unmasked_weight (float, optional): The weight for non-masked cross-entropy loss (Default: ``0.0``).
+        feature_weight (float, optional): The weight for feature penalty loss (Default: ``10.0``).
+        reduction (str, optional): The reduction method for cross-entropy loss (Default: ``"sum"``).
+    """
+    loss = feature_penalty * feature_weight * logit_m.shape[0]
```
[the remark this answers is not in the document]
Comment (nateanl) [reply examples/ssl/losses/_hubert_loss.py:28]: `logit_m` is essential to the training. `logit_u` can be `None`.

File: examples/ssl/train_hubert.py (hunk 1 above)
Comment (nateanl) [reply examples/ssl/train_hubert.py:122]: cool, will change in that way

File: examples/ssl/lightning.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        self.model = model
+        self.loss_fn = loss_fn
+        self.optimizer = optimizer
+        self.lr_scheduler = lr_scheduler
+
+    def configure_optimizers(self):
+        pass
+
+    def log_metric(self, batch, output, loss, step_type):
+        pass
```
[the remark this answers is not in the document]
Comment (hwangjeff) [reply examples/ssl/lightning.py:27]: `configure_optimizers` is defined in `LightningModule`, so it seems unnecessary to override it with an empty function. for `log_metric`, can we perhaps add a comment that explains what we expect users to do with it? also, i'd rename to `log_metrics`

File: examples/ssl/lightning.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+    def training_step(self, batch, batch_idx):
+        opt = self.optimizers()
+        opt.zero_grad()
```
[the remark this answers is not in the document]
Comment (nateanl) [reply examples/ssl/lightning.py]: yes, I forgot to remove it and amp setting. Removed them in bdffc7f0b6e319292e90418cd3a1775b21d28eab

File: examples/ssl/lightning.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        self.optimizer = optimizer
+        self.lr_scheduler = lr_scheduler
+
+    def configure_optimizers(self):
+        pass
+
+    def log_metric(self, batch, output, loss, step_type):
+        pass
+
+    def training_step(self, batch, batch_idx):
```
[the remark this answers is not in the document]
Comment (nateanl) [reply examples/ssl/lightning.py]: resolved in bdffc7f0b6e319292e90418cd3a1775b21d28eab

File: examples/ssl/lightning.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    def log_metrics(self, batch: Batch, output, loss, step_type):
+        """Log useful information to TensorBoard. Users are expected to
+        write their customized `log_metrics` method to log information
+        such as loss values, metric scores, etc.
+
+        Args:
+            batch (Batch): Batch tuple from the dataloader.
+            output: Output generated by the model.
+            loss (Tensor): Generated class
+            step_type (str): Type of step. Choices are "train", "valid", and "test".
```
Comment (hwangjeff) [inline examples/ssl/lightning.py:36]: `validation_step` actually provides "val" rather than "valid"

**Review by hwangjeff — APPROVED** [review]
